pyccel/ast/numpyext.py

```diff
@@ -647,15 +647,22 @@ class NumpyArray(NumpyNewArray):
 
     def __init__(self, arg, dtype=None, order='K', ndmin=None):
 
-        if not isinstance(arg, (PythonTuple, PythonList, Variable)):
+        if not isinstance(arg, (PythonTuple, PythonList, Variable, IndexedElement)):
             raise TypeError('Unknown type of  %s.' % type(arg))
 
         is_homogeneous_tuple = isinstance(arg.class_type, NativeHomogeneousTuple)
-        is_array = isinstance(arg, Variable) and arg.is_ndarray
+        # Inhomogeneous tuples can contain homogeneous data if it is inhomogeneous due to pointers
```

Comment:
Could you please give me an example of what you mean here?

Reply:
```python
a = np.array([1,2,3])
b = np.array([4,5,6])
c = (a,b)
```
`c` contains homogeneous data as all elements have the same dtype, precision, rank, shape, order, class_type. However if we save the variable `c` as a homogeneous tuple then it would have the type:
```fortran
real(F64) :: c(:,:)
```
But then the pointers are lost. If we modify `c[0]` then `a` should also be modified. As there is no such thing as a list of pointers in Fortran the simplest way to handle this is to simply make `c` inhomogeneous so it has the type:
```fortran
real(F64), pointer :: c_0(:)
real(F64), pointer :: c_1(:)
```

Reply:
This makes perfect sense, thanks.

Just a thought. Now that we can handle custom classes we might want to consider another approach in Fortran, where a list of pointers is mimicked by an array of data structures which contain a pointer array:
```fortran
type pntr_to_f64_array
    real(f64), pointer :: p(:)
end type

...

real(f64), allocatable  :: a(:)
real(f64), allocatable  :: b(:)
type(pntr_to_f64_array) :: c(2)

...

c(0) => a
c(1) => b
```
This may be useful for large tuples.

Reply:
This is a good idea. I suggest that we open an issue for this as a later improvement.

That said, the difficulty when creating classes is making them compatible between files. If fileA and fileB both create a homogeneous tuple of pointers then can these objects be passed between functions of the 2 files? With the naive approach where the class is defined in the file this would not be possible as the objects will have different types